core/src/main/java/hudson/model/AbstractBuild.java

```
@@ -459,49 +459,90 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
-                return Result.FAILURE;
+                if (!preBuild(listener,project.getProperties()))
+                    return Result.FAILURE;
 
-            Result result = doRun(listener);
+                result = doRun(listener);
+            } finally {
+                if (!tearDownMarker.tornDown) {
+                    // looks like environments are not torn down yet, do it now (in reverse order)
+                    boolean tearDownFailed = false;
+                    boolean tearDownInterrupted = false;
+                    for (int i = buildEnvironments.size() - 1; i >= 0; i--) {
+                        final Environment environment = buildEnvironments.get(i);
+                        try {
+                            if (!environment.tearDown(AbstractBuild.this, listener)) {
+                                tearDownFailed = true;
+                            }
+                        } catch (IOException | InterruptedException | RuntimeException e) {
+                            tearDownFailed = true;
+                            // exceptions are ignored to give a chance to all environments to tear down
+                            listener.error("Unable to tear down: " + e.getMessage());
+                            Functions.printStackTrace(e, listener.getLogger());
+                            if (e instanceof InterruptedException) {
+                                // don't forget we've been interrupted
+                                tearDownInterrupted = true;
+                            }
+                        }
+                    }
+                    // report any error while tearing down an Environment as a build failure
+                    if (tearDownFailed) {
+                        result = Result.FAILURE;
+                    }
+                    if (tearDownInterrupted) {
+                        // don't forget we've been interrupted
+                        Thread.currentThread().interrupt();
```

Isn't it better to throw an interrupted exception here?

---

Yes, see my answer to your previous comment, and 625b783b9f.